src/main/kotlin/org/rust/lang/core/psi/ext/RsQualifiedNamedElement.kt

```diff
@@ -32,6 +32,27 @@ val RsQualifiedNamedElement.qualifiedName: String?
         return "$cargoTarget$inCratePath"
     }
 
+fun RsQualifiedNamedElement.qualifiedNameRelativeTo(mod: RsMod): String? {
+    val crateRelativePath = crateRelativePath
+    if (mod.crateRoot != crateRoot || crateRelativePath == null) {
+        return qualifiedName
+    }
+
+    val modPathPrefix = mod.crateRelativePath?.let { "$it::" }
+    return if (containingMod.superMods.contains(mod)
+        && modPathPrefix != null
+        && crateRelativePath.startsWith(modPathPrefix)
+    ) {
+        var result = crateRelativePath.removePrefix(modPathPrefix)
+        if (cargoWorkspace?.packages?.any { result.startsWith("${it.normName}::") } == true) {
+            result = "self::$result"
+        }
+        result
+    } else {
+        "crate$crateRelativePath"
```

Comment:
Do I understand correctly that `crateRelativePath` **always** starts with `::`?

Reply:
`crateRelativePath` is empty for crate root (so path would be `crate`, which is correct). Otherwise it always starts with `;;`, because all implementations call [`RsPsiImplUtil::crateRelativePath`](https://github.com/intellij-rust/intellij-rust/blob/eeec35c110b601b6640fab0acf2a859bc08d9305/src/main/kotlin/org/rust/lang/core/psi/RsPsiImplUtil.kt#L21) method, which uses [`RsPsiImplUtil::modCrateRelativePath`](https://github.com/intellij-rust/intellij-rust/blob/eeec35c110b601b6640fab0acf2a859bc08d9305/src/main/kotlin/org/rust/lang/core/psi/RsPsiImplUtil.kt#L27) method, which can return only `null`, empty string, or string prefixed with `::`

I added `check(crateRelativePath.isEmpty() || crateRelativePath.startsWith("::"))`